### PR TITLE
SIDE-931 - Add search UI services / sensors

### PR DESCRIPTION
**Pull request by dmorrison-olive**
https://crosschx.atlassian.net/browse/SIDE-931

### CODE REVIEW

**Review by scottkipferOlive — APPROVED** [review]

**Comment by dmorrison-olive** [comment]
I'm trying to add examples for the global and sidebar searches now...

File: ldk/go/examples/search-global/loop/loop.go
Comment (dmorrison-olive) [inline ldk/go/examples/search-global/loop/loop.go:62]: I guess we'll need to (ideally) take this wrapped goroutine out of the examples when the fix you mentioned is in place, @richseviora? Should I leave a TODO comment here?

**Comment by dmorrison-olive** [comment]
I added some example loops for the global search and searchbar.

The global search one works for me, but I can't get the searchbar example loop to trigger. 🤔

File: ldk/go/examples/search-global/loop/loop.go
Comment (richseviora) [reply ldk/go/examples/search-global/loop/loop.go:62]: After some thought, I'm going to leave the Whisper methods as-is. Users will be instructed that the function will be blocking; if they want to close the whisper early they need to cancel the context they're passing in.

So instead I'll be updating the examples so they're all wrapped in a goroutine.

File: ldk/go/examples/search-global/Makefile
Comment (BillyMurrayOlive) [inline ldk/go/examples/search-global/Makefile:1]: I think this should be:
`REPO := github.com/open-olive/loop-development-kit/ldk/go/example/search-global`

File: ldk/go/examples/search-global/loop/loop_test.go
Comment (BillyMurrayOlive) [inline ldk/go/examples/search-global/loop/loop_test.go:13]: I think this test needs to do more. I would like to see it simulate the user entering text in the global search. Then confirming that it gets a whisper in response. 

Take a look at the whisper-confirm example as a guide. I think a staging approach like the one in that example would be good here.
Ex:
Stage 1 - Simulate user searching text
Stage 2 - Confirm whisper is received with correct content

File: ldk/go/examples/search-global/storage.json
Comment (BillyMurrayOlive) [inline ldk/go/examples/search-global/storage.json:1]: This example doesn't use storage. This file is unused and should be removed.

File: ldk/go/examples/search-searchbar/Makefile
Comment (BillyMurrayOlive) [inline ldk/go/examples/search-searchbar/Makefile:1]: Should be:
` REPO := github.com/open-olive/loop-development-kit/ldk/go/example/search-searchbar`

File: ldk/go/examples/search-searchbar/loop/loop.go
Comment (BillyMurrayOlive) [inline ldk/go/examples/search-searchbar/loop/loop.go:17]: I think this should be more specific. 
`ldk.NewLogger("example-search-searchbar")`
or something like that.

File: ldk/go/examples/search-searchbar/loop/loop_test.go
Comment (BillyMurrayOlive) [inline ldk/go/examples/search-searchbar/loop/loop_test.go:13]: Same comment I made for the search-global example. This test needs more testing.
Also, please update the name to something more descriptive. `TestSearchResponse` or something like that.

File: ldk/go/examples/search-searchbar/storage.json
Comment (BillyMurrayOlive) [inline ldk/go/examples/search-searchbar/storage.json:1]: Unused and should be remove.

**Review by BillyMurrayOlive — CHANGES_REQUESTED** [review]
Good work! I made some suggestions for improving the example tests that I would like to see implemented before this is merged.

**Comment by dmorrison-olive** [comment]
Thanks for reviewing, @BillyMurrayOlive! I'll take a look at your feedback.

File: ldk/go/examples/search-global/Makefile
Comment (dmorrison-olive) [reply ldk/go/examples/search-global/Makefile:1]: Whoops - thanks!

Also, `github.com/open-olive/loop-development-kit/ldk/go/example` should be changed to *examples* everywhere, right?

File: ldk/go/examples/search-searchbar/loop/loop.go
Comment (dmorrison-olive) [reply ldk/go/examples/search-searchbar/loop/loop.go:17]: Changing that gives me pause as it's "loop-example" in all the examples right now, and we'd need to change it everywhere plus remember to change it when creating a new example (if we copy an existing one). Thoughts?

Comment (BillyMurrayOlive) [reply ldk/go/examples/search-searchbar/loop/loop.go:17]: That's a good point. I think it should be specific everywhere. We have missed it in all the other examples. You're right that it will make copy pasting more intensive, but I think it is worth the cost to have logs accurately labeled.

Consider the situation where you might have an example build from something you worked on in the past. You go to deploy a new example for testing and see a log. You assume the log is from your new example because you forgot you had the other example already build. Could mislead a person into following the wrong path.

File: ldk/go/examples/search-global/Makefile
Comment (BillyMurrayOlive) [reply ldk/go/examples/search-global/Makefile]: Oh yes. Good catch. It should be plural.

Comment (dmorrison-olive) [reply ldk/go/examples/search-global/Makefile]: Cool - I'll change it everywhere. 👍

File: ldk/go/examples/search-searchbar/loop/loop.go
Comment (dmorrison-olive) [reply ldk/go/examples/search-searchbar/loop/loop.go:17]: Fair enough counterpoint. That could be confusing in the logs. I'll change this and try to change it everywhere else to be specific as well. 👍

File: ldk/go/examples/search-global/loop/loop.go
Comment (dmorrison-olive) [reply ldk/go/examples/search-global/loop/loop.go:62]: Hey @richseviora and @BillyMurrayOlive, I just noticed that when the call to `Whisper().Markdown()` is wrapped in a goroutine for the clipboard example, that the test always passes (I tried putting in some text to make the assertion fail, and it still worked). I assume this is because I need to put in some coordination (or however you say) between the goroutine and the test assertion? Got any ideas about how to do this?

I'm asking because this would probably apply to the tests for the global and searchbar search example loops.

Comment (dmorrison-olive) [reply ldk/go/examples/search-global/loop/loop.go:61]: Nevermind - I finally grokked just using a channel and blocking on it until it's sent something in the fake Markdown() goroutine. 👍

**Comment by dmorrison-olive** [comment]
> I added some example loops for the global search and searchbar in Go.
> 
> The global search one works for me, but I can't get the searchbar example loop to trigger. 🤔

For me now, the global and searchbar searches sporadically work (maybe 1/3 searches) when I use the samples here. I'm not sure why. 🤔

**Review by JonBernardOlive — APPROVED** [review]

**Comment by BillyMurrayOlive** [comment]
> > I added some example loops for the global search and searchbar in Go.
> > The global search one works for me, but I can't get the searchbar example loop to trigger. 🤔
> 
> For me now, the global and searchbar searches sporadically work (maybe 1/3 searches) when I use the samples here. I'm not sure why. 🤔

**Comment by BillyMurrayOlive** [comment]
> > I added some example loops for the global search and searchbar in Go.
> > The global search one works for me, but I can't get the searchbar example loop to trigger. 🤔
> 
> For me now, the global and searchbar searches sporadically work (maybe 1/3 searches) when I use the samples here. I'm not sure why. 🤔

@dmorrison-olive Can you upload what you have so far for the loop example tests.

**Comment by dmorrison-olive** [comment]
> > > I added some example loops for the global search and searchbar in Go.
> > > The global search one works for me, but I can't get the searchbar example loop to trigger. 🤔
> > 
> > 
> > For me now, the global and searchbar searches sporadically work (maybe 1/3 searches) when I use the samples here. I'm not sure why. 🤔
> 
> @dmorrison-olive Can you upload what you have so far for the loop example tests?

@BillyMurrayOlive, I have everything pushed.

File: ldk/go/examples/search-searchbar/loop/loop_test.go
Comment (DrewRepaskyOlive) [inline ldk/go/examples/search-searchbar/loop/loop_test.go:18]: this reads as "listen-search-barf" 😂

Comment (dmorrison-olive) [reply ldk/go/examples/search-searchbar/loop/loop_test.go:18]: I know, right?! haha

**Review by DrewRepaskyOlive — APPROVED** [review]

**Review by BillyMurrayOlive — APPROVED** [review]

**Pull request merged**
